### PR TITLE
[engine]add cls engine

### PR DESCRIPTION
**Pull request by MarsMeng1994**
不同特征准确率测试：
image

速度测试：
测试集：使用了400条测试样例，每条时长
测试方式：固定model infer线程数为1，取耗时的平均值。
image

### CODE REVIEW

File: runtime/engine/cls/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+set(ARCH "mserver_x86_64" CACHE STRING "Target Architecture: 
+android_arm, android_armv7, android_armv8, android_x86, android_x86_64, 
+mserver_x86_64, ubuntu_x86_64,
+ios_armv7, ios_armv7s, ios_armv8, ios_x86_64, ios_x86,
+windows_x86")
+
+set(CMAKE_VERBOSE_MAKEFILE ON)
+
+set(FASTDEPLOY_DIR ${CMAKE_SOURCE_DIR}/fc_patch/fastdeploy)
```
Comment (lizezheng) [inline runtime/engine/cls/CMakeLists.txt]: 这里是把 linux 和 android 的  fd 库都下载了吗？

File: runtime/engine/cls/nnet/cls_interface.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    cls_nnet_conf.high_freq = conf.Read("high_freq", 14000);
+    cls_nnet_conf.dither = conf.Read("dither", 0.0);
+    
+    ppspeech::ClsNnet* cls_model = new ppspeech::ClsNnet();
+    int ret = cls_model->init(cls_nnet_conf);
+    return (void*)cls_model;
+};
+
+int cls_destroy_instance(void* instance){
+    ppspeech::ClsNnet* cls_model = (ppspeech::ClsNnet*)instance;
```
Comment (zh794390558) [inline runtime/engine/cls/nnet/cls_interface.cc]: (ppspeech::ClsNnet*) ——> static_cast<>

File: runtime/engine/cls/nnet/cls_interface.cc [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    cls_nnet_conf.frame_length_ms = conf.Read("frame_length_ms", 32);
+    cls_nnet_conf.frame_shift_ms = conf.Read("frame_shift_ms", 10);
+    cls_nnet_conf.num_bins = conf.Read("num_bins", 64);
+    cls_nnet_conf.low_freq = conf.Read("low_freq", 50);
+    cls_nnet_conf.high_freq = conf.Read("high_freq", 14000);
+    cls_nnet_conf.dither = conf.Read("dither", 0.0);
+    
+    ppspeech::ClsNnet* cls_model = new ppspeech::ClsNnet();
+    int ret = cls_model->init(cls_nnet_conf);
+    return (void*)cls_model;
```
Comment (zh794390558) [inline runtime/engine/cls/nnet/cls_interface.cc]: static_cast

File: runtime/engine/cls/CMakeLists.txt (hunk 1 above)
Comment (MarsMeng1994) [reply runtime/engine/cls/CMakeLists.txt]: 嗯，安卓编译还没弄，目前只是linux编译

File: runtime/engine/cls/nnet/cls_nnet.cc [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    runtime_option.SetModelPath(conf.model_file_path_, conf.param_file_path_, fastdeploy::ModelFormat::PADDLE);
+    runtime_option.UseLiteBackend();
+#endif
+#ifdef USE_PADDLE_INFERENCE_BACKEND
+    runtime_option.SetModelPath(conf.model_file_path_, conf.param_file_path_, fastdeploy::ModelFormat::PADDLE);
+    runtime_option.UsePaddleInferBackend();
+#endif
+    runtime_option.SetCpuThreadNum(conf.num_cpu_thread_);
+    runtime_option.DeletePaddleBackendPass("simplify_with_basic_ops_pass");
+    runtime_ = std::unique_ptr<fastdeploy::Runtime>(new fastdeploy::Runtime());
```
Comment (zh794390558) [inline runtime/engine/cls/nnet/cls_nnet.cc]: make_unique

File: runtime/engine/cls/nnet/cls_nnet.cc [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+    //init dict
+    if (conf.dict_file_path_ != ""){
+        init_dict(conf.dict_file_path_);
+    }
+
+    // init model
+    fastdeploy::RuntimeOption runtime_option;
+
+#ifdef USE_ORT_BACKEND
```
Comment (lizezheng) [inline runtime/engine/cls/nnet/cls_nnet.cc]: 这种有互斥关系的 宏定义最好用 #elif 会规范些

File: runtime/engine/cls/nnet/cls_nnet.cc [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    ss_ << "}";
+
+    if (result_max_len <= ss_.str().size()){
+        printf("result_max_len is short than result len\n");
+    }
+    snprintf(result, result_max_len, "%s", ss_.str().c_str());
+    return 0;
+};
+
+int ClsNnet::model_forward(const float* features, const int num_frames, const int feat_dim, std::vector<float>& model_out){
```
Comment (zh794390558) [inline runtime/engine/cls/nnet/cls_nnet.cc]: 都改成驼峰命名吧。

**Comment by zh794390558** [comment]
> 使用了400条测试样例，每条时长

时长多少？

File: runtime/engine/cls/CMakeLists.txt (hunk 1 above)
Comment (SmileGoat) [reply runtime/engine/cls/CMakeLists.txt]: fastdeploy.cmake: 放入runtime/cmake/
cmake_minimum_required(VERSION 3.14 FATAL_ERROR)

set(ARCH "mserver_x86_64" CACHE STRING "Target Architecture: 
android_arm, android_armv7, android_armv8, android_x86, android_x86_64, 
mserver_x86_64, ubuntu_x86_64, ios_armv7, ios_armv7s, ios_armv8, ios_x86_64, ios_x86,
windows_x86")

set(CMAKE_VERBOSE_MAKEFILE ON)

set(FASTDEPLOY_DIR ${CMAKE_SOURCE_DIR}/fc_patch/fastdeploy)
if(NOT EXISTS ${FASTDEPLOY_DIR}/fastdeploy-linux-x64-1.0.2.tgz)
    exec_program("mkdir -p ${FASTDEPLOY_DIR} && 
        wget https://bj.bcebos.com/fastdeploy/release/cpp/fastdeploy-linux-x64-1.0.2.tgz -P ${FASTDEPLOY_DIR} &&
        tar xzvf ${FASTDEPLOY_DIR}/fastdeploy-linux-x64-1.0.2.tgz -C ${FASTDEPLOY_DIR} &&
        mv ${FASTDEPLOY_DIR}/fastdeploy-linux-x64-1.0.2 ${FASTDEPLOY_DIR}/linux-x64")
endif()

if(NOT EXISTS ${FASTDEPLOY_DIR}/fastdeploy-android-1.0.0-shared.tgz)
    exec_program("mkdir -p ${FASTDEPLOY_DIR} && 
        wget https://bj.bcebos.com/fastdeploy/release/android/fastdeploy-android-1.0.0-shared.tgz -P ${FASTDEPLOY_DIR} &&
        tar xzvf ${FASTDEPLOY_DIR}/fastdeploy-android-1.0.0-shared.tgz -C ${FASTDEPLOY_DIR} &&
        mv ${FASTDEPLOY_DIR}/fastdeploy-android-1.0.0-shared ${FASTDEPLOY_DIR}/android-armv7v8")
endif()

if (ARCH STREQUAL "mserver_x86_64")
    set(FASTDEPLOY_INSTALL_DIR ${FASTDEPLOY_DIR}/linux-x64)
    add_definitions("-DUSE_PADDLE_INFERENCE_BACKEND")
    # add_definitions("-DUSE_ORT_BACKEND")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -msse -msse2")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -msse -msse2 -mavx -O3")
elseif (ARCH STREQUAL "android_armv7")
    set(FASTDEPLOY_INSTALL_DIR ${FASTDEPLOY_DIR}/android-armv7v8)
    add_definitions("-DUSE_PADDLE_LITE_BAKEND")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g -mfloat-abi=softfp -mfpu=vfpv3 -mfpu=neon -fPIC -pie -fPIE")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -g0 -O3 -mfloat-abi=softfp -mfpu=vfpv3 -mfpu=neon -fPIC -pie -fPIE")
endif()

include(${FASTDEPLOY_INSTALL_DIR}/FastDeploy.cmake)
include_directories(${FASTDEPLOY_INCS})

File: runtime/engine/cls/nnet/cls_interface.cc [hunk 7]
```diff
@@ -0,0 +1,59 @@
+#include "nnet/cls_nnet.h"
+#include "common/base/config.h"
+
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/cls_interface.cc]: 文件夹已经是cls了，那么其文件夹下面文件命名 cls是否多余？？？
coding style 仔细看看，function，变量命名，大小写请重新整理。

File: runtime/engine/common/base/config.h [hunk 8]
```diff
@@ -0,0 +1,352 @@
+#include <iostream>
+#include <fstream>
+#include <map>
+#include <string>
+#include <sstream> 
+using namespace std;
+
+#pragma once
+
+#pragma region ParseIniFile
```
Comment (SmileGoat) [inline runtime/engine/common/base/config.h:13]: ？建议去掉，or 统一一下config

File: runtime/engine/common/base/timer.h [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
```
Comment (SmileGoat) [inline runtime/engine/common/base/timer.h]: 去掉，use kaldi timer；

File: runtime/engine/cls/nnet/wav.h [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef FRONTEND_WAV_H_
+#define FRONTEND_WAV_H_
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/wav.h]: 去掉，复用已有文件

**Review by zh794390558 — COMMENTED** [review]
pre-commit下代码

File: runtime/engine/cls/nnet/panns_interface.cc [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    cls_nnet_conf.num_cpu_thread_ = conf.Read("num_cpu_thread", 12);
+    cls_nnet_conf.samp_freq = conf.Read("samp_freq", 32000);
+    cls_nnet_conf.frame_length_ms = conf.Read("frame_length_ms", 32);
+    cls_nnet_conf.frame_shift_ms = conf.Read("frame_shift_ms", 10);
+    cls_nnet_conf.num_bins = conf.Read("num_bins", 64);
+    cls_nnet_conf.low_freq = conf.Read("low_freq", 50);
+    cls_nnet_conf.high_freq = conf.Read("high_freq", 14000);
+    cls_nnet_conf.dither = conf.Read("dither", 0.0);
+
+    ppspeech::ClsNnet* cls_model = new ppspeech::ClsNnet();
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/panns_interface.cc:41]: 裸指针禁用

File: runtime/engine/cls/nnet/panns_nnet.cc [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+int ClsNnet::InitDict(const std::string& dict_path) {
+    std::ifstream fp(dict_path);
+    std::string line = "";
+    while (getline(fp, line)) {
+        dict_.push_back(line);
+    }
+    return 0;
+}
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/panns_nnet.cc]: 有现成的api使用，见string util or file util

File: runtime/engine/cls/nnet/panns_nnet.cc [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+    timer.Reset();
+#endif
+    // read wav
+    std::ifstream infile(wav_path, std::ifstream::in);
+    kaldi::WaveData wave_data;
+    wave_data.Read(infile);
+    kaldi::Matrix<float> wavform_kaldi = wave_data.Data();
+    int wavform_len = wavform_kaldi.NumRows() * wavform_kaldi.NumCols();
+    std::vector<float> wavform(wavform_kaldi.Data(),
+                               wavform_kaldi.Data() + wavform_len);
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/panns_nnet.cc:102]: 双通道？

File: runtime/engine/cls/nnet/panns_nnet.h [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+  private:
+    int InitDict(const std::string& dict_path);
+    int ModelForward(float* features,
+                     const int num_frames,
+                     const int feat_dim,
+                     std::vector<float>* model_out);
+    int ModelForwardStream(std::vector<float>* feats);
+    int GetTopk(int k, std::vector<float>* model_out);
+    int WaveformFloatNormal(std::vector<float>* waveform);
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/panns_nnet.h]: nnet中的函数为什么有 audio process的功能？这些normal，powerdb，抽出来放到 util中

File: runtime/engine/cls/nnet/panns_nnet_main.cc [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    const char* conf_path = argv[1];
+    const char* wav_path = argv[2];
+    int topk = std::atoi(argv[3]);
+    void* instance = ppspeech::ClsCreateInstance(conf_path);
+    int ret = 0;
+    // read wav
+    std::ifstream ifs(wav_path);
+    std::string line = "";
+    while (getline(ifs, line)) {
+        // read wav
```
Comment (SmileGoat) [inline runtime/engine/cls/nnet/panns_nnet_main.cc:40]: main refactor

Comment (MarsMeng1994) [reply runtime/engine/cls/nnet/panns_nnet_main.cc:40]: done

**Comment by zh794390558** [comment]
处理完的点击下 resolve

**Review by SmileGoat — APPROVED** [review]
LGTM

**Pull request merged**